hax/hax/ha/handler/node.py

```diff
@@ -0,0 +1,64 @@
+# Copyright (c) 2021 Seagate Technology LLC and/or its Affiliates
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+import logging
+
+from hax.ha.events import Event
+from hax.ha.handler import EventHandler
+from hax.message import BroadcastHAStates
+from hax.motr.planner import WorkPlanner
+from hax.types import HAState, ServiceHealth
+from hax.util import ConsulUtil
+
+LOG = logging.getLogger('hax')
+
+
+class NodeEventHandler(EventHandler):
+    """
+    Handles HA events with resource_type == 'node'.
+
+    As a result of such HA event the corresponding BroadcastHAStates is
+    offered to WorkPlanner.
+    """
+    def __init__(self, consul: ConsulUtil, planner: WorkPlanner):
+        """Constructor."""
+        self.cns = consul
+        self.planner = planner
+
+    def handle(self, msg: Event) -> None:
+        node_fid = self.cns.get_node_fid(msg.node_id)
+        if not node_fid:
+            LOG.warn('Unknown [node_id=%s] provided. HA event is ignored',
+                     msg.node_id)
+            return
+
+        get_health = self._get_status_by_text
+
+        self.planner.add_command(
+            BroadcastHAStates(states=[
```

Comment:
This will cause 'broadcast_ha_states' to execute right?
If yes then in that function we need to check if node failure is getting broadcasted(st.fid.container == ObjT.NODE.value) and we should call 'notify_node_status'.
Please correct me if I am wrong.

Reply:
@SwapnilGaonkar7 according to what I see in the code, it doesn't make much sense.

```python
    def notify_node_status(self,
                           proc_note: HaNoteStruct) -> List[HaNoteStruct]:
        new_state = proc_note.no_state
        proc_fid = Fid.from_struct(proc_note.no_id)
        assert ObjT.PROCESS.value == proc_fid.container
        LOG.debug('Notifying node status for process_fid=%s state=%s',
                  proc_fid, new_state)

        node = self.consul_util.get_process_node(proc_fid)

        node_fid = self.consul_util.get_node_fid(node)
        encl_fid = self.consul_util.get_node_encl_fid(node)
        ctrl_fids = self.consul_util.get_node_ctrl_fids(node)
        LOG.debug('node_fid: %s encl_fid: %s ctrl_fids: %s with state: %s',
                  node_fid, encl_fid, ctrl_fids, new_state)

        notes = []
        if node_fid and encl_fid:
            notes = [HaNoteStruct(no_id=x.to_c(), no_state=new_state)
                     for x in [node_fid, encl_fid]]

        if ctrl_fids:
            for x in ctrl_fids:
                notes.append(HaNoteStruct(no_id=x.to_c(), no_state=new_state))

        return notes
```

Facts:
1. This function takes a PROCESS fid, learns the node where this process runs and adds the node state to the list of `HaNoteSruct` to report to Motr layer.
2. The function doesn't update node status in Consul KV anyway.

In my case I don't have a process. Instead, I have the node FID already so this FID is included into the HaNoteStructs that will be broadcast. Do you think I should do anything else?

Reply:
notify_node_status reports encl, ctrls status as failed(We are calling it when all IO services are down)
I think it should also be called when we receive node failure event.

Reply:
> This function takes a PROCESS fid, learns the node where this process runs and adds the node state to the list of HaNoteSruct to report to Motr layer.

Correct, and in case of the node failure event, we won't receive any process failure events. So which process fid should we pass here? Maybe we need to update the function params to receive the node and state.

Reply:
Yes, I'll need to refactor `notify_node_status` function then. Working on that..

Reply:
> The notify_node_status() doesn't update node status in Consul KV anyway.
as part of EOS-23005, I am adding update statuses for node, enclosure, and controllers in consul KV, in this function.

Reply:
Looks good

Reply:
Addressed